sbat.c

```
@@ -120,4 +123,161 @@ EFI_STATUS parse_sbat(char *sbat_base, size_t sbat_size, char *buffer,
+				   badly parsed component_generation will be treated as zero 
+				*/
+				UINT16 sbat_gen = atoi(entry->component_generation);
+				UINT16 sbat_var_gen = atoi(sbat_var_entry->component_generation);
+				if ( sbat_gen < sbat_var_gen ) {
+					dprint(L"component's %a generation: %d. Conflicts with SBAT variable generation %d\n",
+					       entry->component_name, sbat_gen, sbat_var_gen );
+					LogError(L"image did not pass SBAT verification\n");
+					return EFI_SECURITY_VIOLATION;
+				}
+			}
+		}
+	}
+	dprint(L"all entries from SBAT section verified\n");
+	return EFI_SUCCESS;
+}
+
+static BOOLEAN
+is_utf8_bom(CHAR8 *buf, size_t bufsize)
+{
+	unsigned char bom[] = { 0xEF, 0xBB, 0xBF };
+	return !CompareMem(buf, bom, MIN(sizeof(bom), bufsize));
+}
+
+static struct sbat_var *
+new_entry(const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new_entry = AllocatePool(sizeof(*new_entry));
+
+	INIT_LIST_HEAD(&new_entry->list); 
+	new_entry->component_generation = comp_gen;
+	new_entry->component_name_size = comp_name_size;
+	new_entry->component_name = comp_name;
+	return new_entry;
+}
+
+static int
+add_entry(list_t *list, const CHAR8 *comp_gen, const CHAR8 *comp_name_size,
+          const CHAR8 *comp_name)
+{
+	struct sbat_var *new;
+	new = new_entry(comp_gen, comp_name_size, comp_name);
+	if (!new)
+		return -1;
+
+	list_add_tail(&new->list, list);
+	return 0;
+}
+
+static void
+clean_up_vars(list_t *entries)
+{
+	list_t *pos = NULL, *tmp = NULL;
+	for_each_sbat_var_safe(pos, tmp, entries) {
+		struct sbat_var *entry;
+
+		entry = list_entry(pos, struct sbat_var, list);
+		list_del(&entry->list);
+		if (entry->component_generation)
+			FreePool((CHAR8 *)entry->component_generation);
+		if (entry->component_name_size)
+			FreePool((CHAR8 *)entry->component_name_size);
+		if (entry->component_name)
+			FreePool((CHAR8 *)entry->component_name);
+		FreePool(entry);
+	}
+}
+
+int
```

I would use `EFI_STATUS` instead here. That way you could propagate the errors returned by EFI services calls and also would be consistent with other functions in this file.

---

Yeah, that's probably true - my bad for suggesting int instead of EFI_STATUS.